File: src/leetcode_grind/day_587.rs
// https://leetcode.com/problems/minimum-number-of-k-consecutive-bit-flips/description
pub fn min_k_bit_flips(nums: Vec<i32>, k: i32) -> i32 {
    let n = nums.len();
    let mut flipped = vec![false; n];
    let mut valid_flips_from_past_window = 0;
    let mut flip_count = 0;
    let k = k as usize;

    for i in 0..n {
        if i >= k {
            if flipped[i - k] {
                valid_flips_from_past_window -= 1;
            }
        }

        if valid_flips_from_past_window % 2 == nums[i] {
            if i + k > n {
                return -1;
            }
            valid_flips_from_past_window += 1;
            flipped[i] = true;
            flip_count += 1;
        }
    }
    flip_count
}

File: java/leetcode_grind/Day587.java
package leetcode_grind;

import java.util.ArrayDeque;

public class Day587 {
    // https://leetcode.com/problems/minimum-number-of-k-consecutive-bit-flips/description/
    static class Solution1 {
        public int minKBitFlips(int[] nums, int k) {
            var deq = new ArrayDeque<Integer>();
            var flips = 0;
            for (int index = 0; index < nums.length; index++) {
                if (!deq.isEmpty() && deq.peekFirst() < index)
                    deq.removeFirst();

                if ((nums[index] + deq.size()) % 2 == 0) {
                    if (index + k - 1 >= nums.length)
                        return -1;
                    deq.addLast(index + k - 1);
                    flips++;
                }
            }
            return flips;
        }
    }

    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    static class Solution2 {
        int sum = 0;

        public TreeNode bstToGst(TreeNode root) {
            if (root == null) {
                return null;
            }
            bstToGst(root.right);
            var t = root.val;
            root.val += sum;
            sum += t;
            bstToGst(root.left);
            return root;
        }
    }

    static class Solution3 {
        public TreeNode bstToGst(TreeNode root) {
            int sum = 0;
            TreeNode node = root;
            while (node != null) {
                if (node.right == null) {
                    sum += node.val;
                    node.val = sum;
                    node = node.left;
                } else {
                    TreeNode succ = getSuccessor(node);
                    if (succ.left == null) {
                        succ.left = node;
                        node = node.right;
                    } else {
                        succ.left = null;
                        sum += node.val;
                        node.val = sum;
                        node = node.left;
                    }
                }
            }
            return root;
        }

        TreeNode getSuccessor(TreeNode node) {
            TreeNode succ = node.right;
            while (succ.left != null && succ.left != node) {
                succ = succ.left;
            }
            return succ;
        }
    }
}
